src/natcap/invest/stormwater.py

```diff
@@ -0,0 +1,1336 @@
+
+    # (Optional) Calculate stormwater infiltration ratio and volume from
+    # LULC, soil groups, biophysical table, and precipitation
+    if 'ir_a' in list(next(iter(biophysical_dict.values()))):
+        LOGGER.info('Infiltration data detected in biophysical table. '
+            'Will calculate infiltration ratio and volume rasters.')
+        infiltration_ratio_array = numpy.array([
+            [biophysical_dict[lucode][f'ir_{soil_group}'] 
+                for soil_group in ['a', 'b', 'c', 'd']
+            ] for lucode in sorted_lucodes
+        ])
+        infiltration_ratio_task = task_graph.add_task(
+            func=pygeoprocessing.raster_calculator,
+            args=([
+                    (FILES['lulc_aligned_path'], 1), 
+                    (FILES['soil_group_aligned_path'], 1), 
+                    (infiltration_ratio_array, 'raw'), 
+                    (sorted_lucodes, 'raw')], 
+                ratio_op, 
+                FILES['infiltration_ratio_path'], 
+                gdal.GDT_Float32, 
+                NODATA),
+            target_path_list=[FILES['infiltration_ratio_path']],
+            dependent_task_list=[align_task],
+            task_name='calculate stormwater infiltration ratio'
+        )
+        infiltration_volume_task = task_graph.add_task(
+            func=pygeoprocessing.raster_calculator,
+            args=([
+                    (FILES['infiltration_ratio_path'], 1), 
+                    (FILES['precipitation_aligned_path'], 1),
+                    (precipitation_nodata, 'raw'),
+                    (pixel_area, 'raw')],
+                volume_op, 
+                FILES['infiltration_volume_path'], 
+                gdal.GDT_Float32, 
+                NODATA),
+            target_path_list=[FILES['infiltration_volume_path']],
+            dependent_task_list=[align_task, infiltration_ratio_task],
+            task_name='calculate stormwater retention volume'
+        )
+        aggregation_dependencies.append(infiltration_volume_task)
+        data_to_aggregate.append(
+            (FILES['infiltration_ratio_path'], 'IR_mean', 'mean'))
+        data_to_aggregate.append(
+            (FILES['infiltration_volume_path'], 'IV_sum', 'sum'))
+
+    # get all EMC columns from an arbitrary row in the dictionary
+    # strip the first four characters off 'EMC_pollutant' to get pollutant name
+    emc_columns = [key for key in next(iter(biophysical_dict.values()))
+        if key.startswith('emc_')]
+    pollutants = [key[4:] for key in  emc_columns]
+    LOGGER.info(f'Pollutants found in biophysical table: {pollutants}')
+    
+    # Calculate avoided pollutant load for each pollutant from retention volume
+    # and biophysical table EMC value
+    avoided_load_paths = []
+    for pollutant in pollutants:
+        # one output raster for each pollutant
+        avoided_pollutant_load_path = os.path.join(
+            output_dir, f'avoided_pollutant_load_{pollutant}{suffix}.tif')
```

Comment:
I think the full path still should be defined in the loop here because it contains the `pollutant` variable. To put them into the file registry you'd have to give each one a key with the variable name in it as well: `{f'{pollutant}_load_path': f'avoided_pollutant_load_{pollutant}{suffix}.tif'...}` Is that what you meant?